prog/dftb+/lib_dftb/densitymatrix.F90

```diff
@@ -706,4 +714,145 @@ subroutine sp_energy_density_matrix_cmplx(dm, eigenvecs, filling, eigen, iNeighb
 
   end subroutine sp_energy_density_matrix_cmplx
 
+
+#:if WITH_SCALAPACK
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+!!! Scalapack routines
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+  !> Create energy weighted density matrix (real).
+  !!
+  subroutine makeDensityMtxRealBlacs(myBlacs, desc, filling, eigenVecs, densityMtx, eigenVals)
+    type(blacsgrid), intent(in) :: myBlacs
+    integer, intent(in) :: desc(:)
+    real(dp), target, intent(in) :: filling(:)
+    real(dp), intent(inout) :: eigenVecs(:,:)
+    real(dp), intent(out) :: densityMtx(:,:)
+    real(dp), intent(in), optional :: eigenVals(:)
+
+    integer  :: ii, jj, iGlob, iLoc, blockSize, nLevel
+    real(dp) :: minFill, maxFill, alpha, beta
+    real(dp), allocatable, target :: eFilling(:)
+    real(dp), pointer :: myFilling(:)
+    type(blocklist) :: blocks
+
+    densityMtx(:,:) = 0.0_dp
+
+    ! Make non-zero fillings positive definite
+    nLevel = size(filling)
+    do while (abs(filling(nLevel)) < epsilon(1.0_dp) .and. nLevel > 1)
+      nLevel = nLevel - 1
+    end do
+    if (present(eigenVals)) then
+      allocate(eFilling(nLevel))
+      eFilling(:) = filling(1:nLevel) * eigenVals(1:nLevel)
+      myFilling => eFilling
+    else
+      myFilling => filling
+    end if
+    minFill = minval(myFilling)
+    maxFill = maxval(myFilling)
+    if ((minFill < 0.0_dp .eqv. maxFill < 0.0_dp) .and. abs(minFill) >= epsilon(1.0_dp)&
+        & .and. abs(maxFill) >= epsilon(1.0_dp)) then
+      alpha = sign(1.0_dp, maxFill)
+      beta = 0.0_dp
+    else
+      alpha = 1.0_dp
+      beta = minFill - arbitraryConstant
+      call pblasfx_psyrk(eigenVecs, desc, densityMtx, desc, kk=nLevel)
+    end if
+
+    ! Scale eigenvectors
```

Comment:
If enough memory, could bypass having to scale and then unscale vectors by using a local copy of the block of the distributed array

Reply:
It mimics, how we do it in serial. But I agree, we should think about making it cleaner.